File: src/game/GameSession.ts
import { GameState } from "../engine/core/GameState.js";

const MIN_RENDER_DISTANCE = 2;
const MAX_RENDER_DISTANCE = 32;

export class GameSession {
  private currentState = GameState.BOOTING;
  private currentRenderDistance: number;

  constructor(renderDistance: number) {
    this.currentRenderDistance = this.clampRenderDistance(renderDistance);
  }

  get state(): GameState {
    return this.currentState;
  }

  get renderDistance(): number {
    return this.currentRenderDistance;
  }

  enterMainMenu(): void {
    this.currentState = GameState.MAIN_MENU;
  }

  startSingleplayer(): void {
    this.currentState = GameState.GENERATING_WORLD;
  }

  markWorldReady(): boolean {
    if (this.currentState !== GameState.GENERATING_WORLD) return false;
    this.currentState = GameState.IN_GAME;
    return true;
  }

  pause(): boolean {
    if (this.currentState !== GameState.IN_GAME) return false;
    this.currentState = GameState.PAUSED;
    return true;
  }

  resume(): boolean {
    if (this.currentState !== GameState.PAUSED) return false;
    this.currentState = GameState.IN_GAME;
    return true;
  }

  returnToTitle(): void {
    this.currentState = GameState.MAIN_MENU;
  }

  setRenderDistance(renderDistance: number): void {
    this.currentRenderDistance = this.clampRenderDistance(renderDistance);
  }

  private clampRenderDistance(renderDistance: number): number {
    const normalized = Number.isFinite(renderDistance) ? Math.floor(renderDistance) : MIN_RENDER_DISTANCE;
    return Math.max(MIN_RENDER_DISTANCE, Math.min(MAX_RENDER_DISTANCE, normalized));
  }
}

File: src/game/GameSession.cpp
#include "GameSession.hpp"
#include "engine/core/RenderDistanceLimits.hpp"
#include <algorithm>

namespace terrain {

GameSession::GameSession(int32_t renderDistance)
  : m_renderDistance(clampRenderDistance(renderDistance)) {}

void GameSession::enterMainMenu() { m_state = GameState::MainMenu; }

void GameSession::startSingleplayer(GameMode mode) {
  m_gameMode = mode;
  m_state = GameState::GeneratingWorld;
  ++m_startRequestId;
}

auto GameSession::markWorldReady() -> bool {
  if (m_state != GameState::GeneratingWorld) return false;
  m_state = GameState::InGame;
  return true;
}

auto GameSession::pause() -> bool {
  if (m_state != GameState::InGame) return false;
  m_state = GameState::Paused;
  return true;
}

auto GameSession::resume() -> bool {
  if (m_state != GameState::Paused) return false;
  m_state = GameState::InGame;
  return true;
}

void GameSession::returnToTitle() { m_state = GameState::MainMenu; }

void GameSession::setRenderDistance(int32_t rd) { m_renderDistance = clampRenderDistance(rd); }

auto GameSession::clampRenderDistance(int32_t rd) -> int32_t {
  return std::clamp(rd, MIN_RENDER_DISTANCE, MAX_RENDER_DISTANCE);
}

} // namespace terrain
